src/ol/render/webgl/VectorStyleRenderer.js

```diff
@@ -0,0 +1,440 @@
+/**
+ * @module ol/render/webgl/VectorStyleRenderer
+ */
+import WebGLArrayBuffer from '../../webgl/Buffer.js';
+import {ARRAY_BUFFER, DYNAMIC_DRAW, ELEMENT_ARRAY_BUFFER} from '../../webgl.js';
+import {AttributeType} from '../../webgl/Helper.js';
+import {WebGLWorkerMessageType} from './constants.js';
+import {
+  create as createTransform,
+  makeInverse as makeInverseTransform,
+} from '../../transform.js';
+import {create as createWebGLWorker} from '../../worker/webgl.js';
+import {
+  generateLineStringRenderInstructions,
+  generatePointRenderInstructions,
+  generatePolygonRenderInstructions,
+  getCustomAttributesSize,
+} from './renderinstructions.js';
+import {parseLiteralStyle} from '../../webgl/styleparser.js';
+
+const WEBGL_WORKER = createWebGLWorker();
+let workerMessageCounter = 0;
+
+/**
+ * Names of attributes made available to the vertex shader.
+ * Please note: changing these *will* break custom shaders!
```

Comment:
Small note that probably deserves a longer discussion elsewhere:  I don't think we should try to support custom shaders.  If we did allow people to provide arbitrary shaders, it would impose some serious constraints on how we could continue evolving the rendering implementation.  We likely couldn't do things like upgrade from WebGL 1 to 2, or migrate to WebGPU, for example.

Reply:
Good point, I realize now that my intentions were more to offer the possibility to use the ShaderBuilder instead of a literal style, and then be able to write custom expressions for color, width, etc. thus offering much more freedom while still providing an abstraction layer on top of GLSL. This could ease things up when moving to WebGL2 for example.

Maybe the ShaderBuilder could be made more usable and become a first class citizen of the API as an alternative to literal styles?